Data/DList/Internal.hs

```diff
@@ -418,40 +391,40 @@ raised.
 @head@ obeys the law:
 
 @
-__head__ ('fromList' (x : xs)) = x
+__head__ xs = 'List.head' ('toList' xs)
 @
 
-Note that @head@ is implemented with 'list'.
-
 -}
 {- ORMOLU_ENABLE -}
 
 {-# INLINE head #-}
 head :: DList a -> a
-head = list (error "Data.DList.head: empty DList") const
+head xs = case toList xs of
+  x : _ -> x
+  [] -> error "Data.DList.head: empty DList"
 
 {- ORMOLU_DISABLE -}
 {-|
 
-__@tail xs@__ is a 'DList' excluding the first element of __@xs@__. If @xs@ is
-empty, an 'error' is raised.
+__@tail xs@__ is a list of the elements in __@xs@__ excluding the first element.
+If @xs@ is empty, an 'error' is raised.
 
 \(\mathcal{O}\)(@'length' ('toList' xs)@).
 
 @tail@ obeys the law:
 
 @
-__tail__ ('fromList' (x : xs)) = xs
+__tail__ xs = 'List.tail' ('toList' xs)
 @
 
-Note that @tail@ is implemented with 'list'.
-
 -}
 {- ORMOLU_ENABLE -}
 
 {-# INLINE tail #-}
-tail :: DList a -> DList a
-tail = list (error "Data.DList.tail: empty DList") (flip const)
+tail :: DList a -> [a]
```

Comment:
@spl 
What speaks against the following definition?
```haskell
tail :: DList a -> DList a
tail (UnsafeDList f) = UnsafeDList (Data.List.tail . f)
```
This would be a proper efficient `tail` function for `DList`.
The new `tail` that you implemented leaves the realm of `DList`s, so it seems it is already subsumed by `Data.List.tail . toList`.